capstone/src/main/java/com/google/sps/servlets/UserGroupsServlet.java

```diff
@@ -47,47 +47,71 @@
 @WebServlet("/user-groups")
 public class UserGroupsServlet extends HttpServlet {
 
+  private UserService userService = UserServiceFactory.getUserService();
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
   /**
    * Gets User's Group data from the Datastore and returns it.
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    UserService userService = UserServiceFactory.getUserService();
     Entity currentUser = null;
     ArrayList<UserGroupResponse> groups = new ArrayList<>();
 
     if (userService.isUserLoggedIn()) {
-      String userId = userService.getCurrentUser().getUserId();
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-      Key entityKey = KeyFactory.createKey("User", userId);
-      try {
-        currentUser = datastore.get(entityKey);
-      } catch (EntityNotFoundException e) {
-        ErrorHandler.sendError(response, "User not found.");
-        return;
-      }
+      currentUser = getUserEntity(response);
 
       LinkedHashSet<Long> groupIds = (currentUser.getProperty("groups") == null)
         ? new LinkedHashSet<>()
         : new LinkedHashSet<Long>((ArrayList<Long>) currentUser.getProperty("groups"));
       
-      for (long groupId : groupIds) {
-        Entity groupEntity = null;
-        try {
-          groupEntity = datastore.get(KeyFactory.createKey("Group", groupId));
-        } catch (EntityNotFoundException e) {
-          ErrorHandler.sendError(response, "Group does not exist.");
-        }
-        UserGroupResponse groupResponse = UserGroupResponse.fromEntity(groupEntity);
-        groups.add(groupResponse);
-      }
+      groups = getGroupsFromIds(groupIds, response);
     } else {
       ErrorHandler.sendError(response, "User not logged in.");
       return;
     }
 
+    // If either of these values are null, an error occured. Do not continue.
+    if (currentUser == null || groups == null) {
+      return;
+    }
```

Comment:
Should this not also send an error instead of just returning early?

Reply:
If either of those are null, then I think an error message would've been sent already from one of the other methods so it would be redundant.